www/blog/2022-11-23-library-ts-perf.md

```diff
@@ -0,0 +1,197 @@
+---
+slug: library-ts-perf
+# TODO: better title (pls help me)
+title: How can you improve the performance of your TypeScript library?
+author: Sachin Raja
+author_title: tRPC Core Team Member
+author_url: https://twitter.com/s4chinraja
+author_image_url: https://avatars1.githubusercontent.com/u/58836760?s=460&v=4
+---
+
+As library authors, we seek to provide the best possible developer experience for our users. Reducing time-to-error, narrowing down exactly what can be passed and accessed where, and providing intuitive APIs at every step all contribute to this. TypeScript has become a major tool in enabling us to ship such DX.
+
+However, as TypeScript usage becomes more widespread and libraries like tRPC and, more recently, [TanStack Router](https://tanstack.com/router/v1) are built entirely around it, we must understand the implications on type-checking and language server performance.
+
+## Where to start?
+
+First, consider if you need to optimize your types. App developers should not have to worry about types on that level, but library maintainers play a crucial role in ensuring their editing experience remains smooth.
+
+Regardless of whether your library is slow _now_, it's important to keep an eye on performance as your library changes. You must test your library's performance to the max **on each commit**.
+
+For tRPC, we ensure this by [generating](https://github.com/trpc/trpc/blob/9fc2d06a8924da73e10b9d4497f3a1f53de706ed/scripts/generateBigBoi.ts) and [testing](https://github.com/trpc/trpc/blob/9fc2d06a8924da73e10b9d4497f3a1f53de706ed/packages/tests/server/react/bigBoi.test.tsx) a router with 3,500 procedures and 1,000 routers. We test all aspects of the library: server, vanilla client, and the React client because they all have different code (type?) paths and have individually regressed in the past.
```

Comment:
I don't think we should use this as a great case, this is more about ensuring it compiles at all but we don't test any performance here.

Reply:
should i just remove it then?